app/attributes/attributecontroller.cpp

```diff
@@ -608,6 +608,41 @@ void AttributeController::recalculateDerivedItems( bool isFormValueChange, bool
     }
   }
 
+  // evaluate if numbers are in correct range
+  {
+    QMap<QUuid, std::shared_ptr<FormItem>>::iterator formItemsIterator = mFormItems.begin();
+    while ( formItemsIterator != mFormItems.end() )
+    {
+      std::shared_ptr<FormItem> item = formItemsIterator.value();
+      QVariant value = featureLayerPair().feature().attribute( item->fieldIndex() );
+
+      bool isRangeEditable = item->editorWidgetType() == QStringLiteral( "Range" ) &&
+                             item->editorWidgetConfig()[QStringLiteral( "Style" )] == QStringLiteral( "SpinBox" );
+
+      if ( isRangeEditable && !value.isNull() )
+      {
+        double min = item->editorWidgetConfig()["Min"].toDouble();
+        double max = item->editorWidgetConfig()["Max"].toDouble();
+        double val = featureLayerPair().feature().attribute( item->fieldIndex() ).toDouble();
+
+        if ( !( min <= val && val <= max ) )
+        {
+          item->setState( FormItem::ValueOutOfRange );
+          changedFormItems << item->id();
+        }
+      }
```

Comment:
best to setup some minimal framework so another validators can be added easily/testable and not mess up the attribute controller logic. 

for example in the folder attributes/ create class fieldvalidator.[h/cpp] a method `FormItem::Validity validate(const std::shared_ptr<FormItem>)` and this implementation to be hidden in this class


OR could be just a static method in rangewidgethelper.h ?

Reply:
https://github.com/lutraconsulting/input/issues/1517